mmcv/cnn/rfsearch/search.py

```diff
@@ -0,0 +1,249 @@
+import logging
+import os
+from typing import Dict
+
+import torch  # noqa
+import torch.nn as nn
+
+from mmcv.runner import HOOKS, Hook
+from .operator import Conv2dRFSearchOp, ConvRFSearchOp  # noqa
+from .utils import load_structure, write_to_json
+
+logging.basicConfig(
+    format='[%(asctime)s-%(filename)s-%(levelname)s:%(message)s]',
+    level=logging.NOTSET,
+    datefmt='%Y-%m-%d%I:%M:%S %p',
+)
+logger = logging.getLogger('Searcher')
+logger.setLevel(logging.ERROR)
+
+
+@HOOKS.register_module()
+class RFSearch(Hook):
+    """Rcecptive field search via dilation rates.
+        Paper: RF-Next: Efficient Receptive Field
+            Search for Convolutional Neural Networks
+
+    Args:
+        logdir (str, optional): save path of searched structure.
+                        Defaults to './log'.
+        mode (str, optional):
+                search/fixed_single_branch/fixed_multi_branch.
+        config (Dict, optional): config dict of search.
+        rfstructure_file (Optional[str], optional):
+                searched recptive fields of the model.
+    """
+
+    def __init__(self,
+                 logdir: str = './log',
+                 mode: str = 'search',
+                 config: Dict = {},
+                 rfstructure_file: str = None):
+        assert logdir is not None
+        assert mode in ['search', 'fixed_single_branch', 'fixed_multi_branch']
+        assert config is not None
+        self.config = config
+        self.config['structure'] = {}
+        if rfstructure_file is not None:
+            rfstructure = load_structure(rfstructure_file)
+            self.config['structure'] = rfstructure
+        self.logdir = logdir
+        self.mode = mode
+        self.S = self.config['search']['S']
+        os.makedirs(self.logdir, exist_ok=True)
+
+    def model_init(self, model: nn.Module):
+        """init model with search ability.
+
+        Args:
+            model (nn.Module): pytorch model
+
+        Raises:
+            NotImplementedError:
+                only support three modes:
+                    search/fixed_single_branch/fixed_multi_branch
+        """
+        print('RFSearch init begin.')
+        # print(runner.model)
+        if self.mode == 'search':
+            if self.config['structure']:
+                self.set_model(model, self.config, search_op='Conv2d')
+            self.wrap_model(model, self.config, search_op='Conv2d')
+        elif self.mode == 'fixed_single_branch':
+            self.set_model(model, self.config, search_op='Conv2d')
+        elif self.mode == 'fixed_multi_branch':
+            self.set_model(model, self.config, search_op='Conv2d')
+            self.wrap_model(model, self.config, search_op='Conv2d')
+        else:
+            raise NotImplementedError
+        print('RFSearch init end.')
+        pass
+
+    def before_run(self, runner):
+        pass
+
+    def after_run(self, runner):
+        pass
+
+    def before_epoch(self, runner):
+        pass
```

Comment:
Those functions can be deleted because they are the same as father class functions.